pypsa/optimization/optimize.py

```diff
@@ -111,83 +170,110 @@ def define_objective(n: Network, sns: pd.Index) -> None:
+            objective.append((operation * cost).sum(dim=["name", "snapshot"]))
+
+    # marginal cost quadratic
+    for c_name, attr in lookup.query("marginal_cost_quadratic").index:
+        c = as_components(n, c_name)
+
+        if c.static.empty or "marginal_cost_quadratic" not in c.static.columns:
+            continue
+
+        cost = c.da.marginal_cost_quadratic.sel(snapshot=sns)
+        if cost.size == 0 or (cost == 0).all():
+            continue
+
+        cost = cost * weight
+
+        operation = m[f"{c.name}-{attr}"].sel(
+            snapshot=sns, name=cost.coords["name"].values
+        )
+        objective.append((operation * operation * cost).sum(dim=["name", "snapshot"]))
+        is_quadratic = True
 
     # stand-by cost
-    comps = {"Generator", "Link"}
-    for c in comps:
-        com_i = get_committable_i(n, c)
+    for c_name in ["Generator", "Link"]:
+        c = as_components(n, c_name)
+        com_i = c.committables
 
         if com_i.empty:
             continue
 
-        stand_by_cost = (
-            get_as_dense(n, c, "stand_by_cost", sns, com_i)
-            .loc[:, lambda ds: (ds != 0).any()]
-            .mul(weighting, axis=0)
+        stand_by_cost = c.da.stand_by_cost.sel(name=com_i, snapshot=sns)
+        if stand_by_cost.size == 0 or (stand_by_cost == 0).all():
+            continue
+
+        stand_by_cost = stand_by_cost * weight
+
+        status = m[f"{c.name}-status"].sel(
+            snapshot=sns, name=stand_by_cost.coords["name"].values
         )
-        stand_by_cost.columns.name = f"{c}-com"
-        status = n.model.variables[f"{c}-status"].loc[:, stand_by_cost.columns]
-        objective.append((status * stand_by_cost).sum())
+        objective.append((status * stand_by_cost).sum(dim=["name", "snapshot"]))
 
     # investment
-    for c, attr in nominal_attrs.items():
-        ext_i = n.get_extendable_i(c)
-        cost = n.static(c)["capital_cost"][ext_i]
-        if cost.empty:
+    for c_name, attr in nominal_attrs.items():
+        c = as_components(n, c_name)
+        ext_i = c.extendables
+
+        if ext_i.empty:
+            continue
+
+        capital_cost = c.da.capital_cost.sel(name=ext_i)
+        if capital_cost.size == 0 or (capital_cost == 0).all():
             continue
 
+        # only charge capex for already-existing assets
         if n._multi_invest:
-            active = pd.concat(
-                {
-                    period: n.get_active_assets(c, period)[ext_i]
-                    for period in sns.unique("period")
-                },
-                axis=1,
-            )
-            cost = active @ period_weighting * cost
+            weighted_cost = 0
+            for i, period in enumerate(periods):
+                # collapse time axis via any() so capex value isn't broadcasted
+                active = c.da.active.sel(period=period, name=ext_i).any(dim="timestep")
+                weighted_cost += capital_cost * active * period_weighting.iloc[i]
```

Comment:
This `.iloc[]` is potentially dangerous, isn't it? E.g. when you solve with multiple investment periods but do not pass all investment periods through the `snapshots` argument (see e.g. https://github.com/PyPSA/PyPSA/pull/1308). Why not a `.loc[period]`?